system/lib/fetch/emscripten_fetch.cpp

```diff
@@ -66,16 +68,65 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+	if (!fetch) return 0;
 	memset(fetch, 0, sizeof(emscripten_fetch_t));
 	fetch->id = globalFetchIdCounter++; // TODO: make this thread-safe!
 	fetch->userData = fetch_attr->userData;
-	fetch->url = strdup(url); // TODO: free
+	fetch->url = strdup(url);
+	if(!fetch->url)
+	{
+		fetch_free(fetch);
+		return 0;
+	}
 	fetch->__attributes = *fetch_attr;
-	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0; // TODO: free
-	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0; // TODO: free
-	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0; // TODO: free
-	fetch->__attributes.requestHeaders = 0;// TODO:strdup(fetch->__attributes.requestHeaders);
-	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0; // TODO: free
+
+#define STRDUP_OR_ABORT(s)				\
+	if (s)								\
+	{									\
+		s = strdup(s);					\
+		if (!s)							\
+		{								\
+			fetch_free(fetch);			\
+			return 0;					\
+		}								\
+	}
+
+	STRDUP_OR_ABORT(fetch->__attributes.destinationPath);
```

Comment:
Hmm, this is not correct. `fetch->__attributes` is assigned from the original source pointers above at `fetch->__attributes = *fetch_attr;`. If the line `STRDUP_OR_ABORT(fetch->__attributes.destinationPath);` here fails allocation, it calls to `fetch_free(fetch);` and as result, it will `free()` all the fields of `fetch->__attributes`, in particular `fetch->__attributes.userName` and `fetch->__attributes.password` - however these pointers have not been strdupped yet, that is only done on the two subsequent lines. So this means that if this first line needs to abort, the function `fetch_free()` will end up freeing user submitted pointers.

Reply:
A good way to make this clean would be to remove the line `fetch->__attributes = *fetch_attr;` above, and instead copy the fields one by one:

```
fetch->__attributes.destinationPath = STRDUP_OR_ABORT(fetch_attr->destinationPath);
```

and have `STRDUP_OR_ABORT(s)` not do a `s = strdup(s);` but instead pass out the dupped pointer?